zipkin-server/src/main/java/zipkin2/server/internal/throttle/ThrottledCall.java

```diff
@@ -0,0 +1,240 @@
+  }
+
+  @Override
+  public V execute() throws IOException {
+    try {
+      call = delegate.get();
+
+      // Make sure we throttle
+      Future<V> future = executor.submit(() -> {
+        String oldName = setCurrentThreadName(call.toString());
+        try {
+          return call.execute();
+        } finally {
+          setCurrentThreadName(oldName);
+        }
+      });
+      V result = future.get(); // Still block for the response
+
+      limitListener.onSuccess();
+      return result;
+    } catch (ExecutionException e) {
+      Throwable cause = e.getCause();
+      if (cause instanceof RejectedExecutionException) {
+        // Storage rejected us, throttle back
+        limitListener.onDropped();
+      } else {
+        limitListener.onIgnore();
+      }
+
+      if (cause instanceof RuntimeException) {
+        throw (RuntimeException) cause;
+      } else if (cause instanceof IOException) {
+        throw (IOException) cause;
+      } else {
+        throw new RuntimeException("Issue while executing on a throttled call", cause);
+      }
+    } catch (InterruptedException e) {
+      limitListener.onIgnore();
+      throw new RuntimeException("Interrupted while blocking on a throttled call", e);
+    } catch (Exception e) {
+      // Ignoring in all cases here because storage itself isn't saying we need to throttle.  Though, we may still be
+      // write bound, but a drop in concurrency won't necessarily help.
+      limitListener.onIgnore();
+      throw e;
+    }
+  }
+
+  @Override
+  public void enqueue(Callback<V> callback) {
+    try {
+      executor.execute(new QueuedCall(callback));
+    } catch (Exception e) {
+      // Ignoring in all cases here because storage itself isn't saying we need to throttle.  Though, we may still be
+      // write bound, but a drop in concurrency won't necessarily help.
+      limitListener.onIgnore();
+      throw e;
+    }
+  }
+
+  @Override
+  public void cancel() {
+    canceled = true;
+    if (call != null) {
+      call.cancel();
+    }
+  }
+
+  @Override
+  public boolean isCanceled() {
+    return canceled || (call != null && call.isCanceled());
+  }
+
+  @Override
+  public Call<V> clone() {
+    return new ThrottledCall<>(this);
+  }
+
+  /**
+   * @param name New name for the current Thread
+   * @return Previous name of the current Thread
+   */
+  static String setCurrentThreadName(String name) {
+    Thread thread = Thread.currentThread();
+    String originalName = thread.getName();
+    try {
+      thread.setName(name);
+      return originalName;
+    } catch (SecurityException e) {
+      return originalName;
+    }
+  }
+
+  final class QueuedCall implements Runnable {
+    final Callback<V> callback;
+
+    public QueuedCall(Callback<V> callback) {
+      this.callback = callback;
+    }
+
+    @Override
+    public void run() {
+      try {
+        if (canceled) {
```

Comment:
do you want the ignore callback invoked on cancel? just asking. maybe comment why this works